src/verdant/panel/events/event-date.tsx

```diff
@@ -25,47 +27,93 @@ type NotebookDate_Props = {
+      accumulator: number[][]; // Holds partially constructed bundle output
+      timeBound: number; // Lower limit on time for inclusion in latest bundle
+      lastType: CheckpointType; // Type of current bundle
+    }
+
+    const reducer = (accObj: accumulatorObject, e: eventState, idx) => {
+      // Compute properties of current element
+      let timeStamp = e.events[0].timestamp;
+      let eventType = this.eventType(e);
+      if ((timeStamp > accObj.timeBound) &&
+        (eventType === accObj.lastType)) { // add event to current bundle
+        const newAccumulator = accObj.accumulator.slice(0, -1).concat(
+          [accObj.accumulator[accObj.accumulator.length - 1].concat([idx])]
+        )
+        return {
+          accumulator: newAccumulator,
+          timeBound: accObj.timeBound,
+          lastType: accObj.lastType
+        }
+      } else { // create new bundle
+        return {
+          accumulator: accObj.accumulator.concat([[idx]]),
+          timeBound: timeStamp - INTERVAL_WIDTH,
+          lastType: eventType
+        }
+      }
     }
+    return events.reduceRight(reducer, {
+      accumulator: [],
+      timeBound: Infinity,
+      lastType: null
+    }).accumulator
+  }
+
+  makeBundles() {
+    const bundledIndices = this.computeBundles(this.props.events);
+
+    // Creates DateBundle for each set of dates
+    return bundledIndices.map(
+      (idx_list, i) => (
+        <div key={i}>
+          <NotebookEventDateBundle
+            bundle_id={i}
+            events={[...idx_list]}
+            date_id={this.props.date_id}
+          />
+        </div>
+      )
+    )
   }
 
   render() {
     return (
       <div>
-        <div
-          className={DATE_HEADER}
-          onClick={() => {
-            if (this.props.isOpen) this.props.close(this.props.date_id);
-            else this.props.open(this.props.date_id);
-          }}
-        >
+        <div className={DATE_HEADER}
+             onClick={() => {
+               if (this.props.isOpen) this.props.close(this.props.date_id);
+               else this.props.open(this.props.date_id);
+             }}>
           <div className={DATE_LABEL}>
             {Checkpoint.formatDate(this.props.date)}
           </div>
           <div className={DATE_COLLAPSE_HEADER}>
             <div
-              className={`${DATE_COUNT} ${this.props.isOpen ? "hidden" : ""}`}
-            >
+              className={`${DATE_COUNT} ${this.props.isOpen ? "hidden" : ""}`}>
               ({this.props.events.length})
             </div>
             <div
-              className={`${DATE_ARROW} ${this.props.isOpen ? "" : "closed"}`}
-            ></div>
+              className={`${DATE_ARROW} ${this.props.isOpen ? "" : "closed"}`}>
+            </div>
           </div>
         </div>
-        <div className={DATE_GROUP}>{this.makeEventList()}</div>
+        <div className={DATE_GROUP}>
+          {this.props.isOpen ? this.makeBundles() : null}
```

Comment:
Are the bundles recorded anywhere in the redux store? Like, if we compute bundles once for this date during the user's session, do we need to re-compute the bundles every time they open and close the date header?

Reply:
They aren't. I thought that because the bundles would be changing for the current date, storing the bundles would get messy. Should I be storing the bundles for all but the current date?

Reply:
Yes, just for prior dates makes sense for now, if you can find a lightweight way of storing which event id's should stay bundled together in the current session. Realistically, we could long-term think about saving bundle ids to disk as a part of the event serialization so that bundles don't have to be computed too often.

Reply:
I added storage of bundles, recomputing only if the bundles don't exist on the dateState or if the number of events and number of events in the bundles are unequal (indicating that events have been added to that date).